motionplan/cBiRRT.go

```diff
@@ -110,115 +105,167 @@ func NewCBiRRTMotionPlannerWithSeed(frame referenceframe.Frame, nCPU int, seed *
-	map1, map2 := seedMap, goalMap
+	map1, map2 := rm.startMap, rm.goalMap
 
-	// TODO(rb) package neighborManager better
-	nm := &neighborManager{nCPU: mp.nCPU}
-	nmContext, cancel := context.WithCancel(ctx)
-	defer cancel()
+	m1chan := make(chan node, 1)
+	m2chan := make(chan node, 1)
+	defer close(m1chan)
+	defer close(m2chan)
 
-	// main sampling loop - for the first sample we try the 0.5 interpolation between seed and goal[0]
-	target := referenceframe.InterpolateInputs(seed, solutions[0].Q(), 0.5)
 	for i := 0; i < algOpts.PlanIter; i++ {
+		if time.Since(mp.start) > time.Duration(algOpts.Timeout)*time.Second {
+			solutionChan <- &rrtPlanReturn{err: errPlannerTimeout, rm: rm}
+			return
+		}
+
 		select {
 		case <-ctx.Done():
-			solutionChan <- &planReturn{err: ctx.Err()}
+			solutionChan <- &rrtPlanReturn{err: ctx.Err()}
 			return
 		default:
 		}
 
-		// attempt to extend map1 first
+		// attempt to extend maps 1 and 2 towards the target
 		nearest1 := nm.nearestNeighbor(nmContext, planOpts, target, map1)
-		map1reached := mp.constrainedExtend(ctx, algOpts, map1, nearest1, &basicNode{q: target})
-
-		// then attempt to extend map2 towards map 1
-		nearest2 := nm.nearestNeighbor(nmContext, planOpts, map1reached.Q(), map2)
-		map2reached := mp.constrainedExtend(ctx, algOpts, map2, nearest2, map1reached)
+		nearest2 := nm.nearestNeighbor(nmContext, planOpts, target, map2)
+		//nolint: gosec
+		rseed1 := rand.New(rand.NewSource(int64(mp.randseed.Int())))
+		//nolint: gosec
+		rseed2 := rand.New(rand.NewSource(int64(mp.randseed.Int())))
+
+		utils.PanicCapturingGo(func() {
+			mp.constrainedExtend(ctx, algOpts, rseed1, map1, nearest1, &basicNode{q: target}, m1chan)
+		})
+		utils.PanicCapturingGo(func() {
+			mp.constrainedExtend(ctx, algOpts, rseed2, map2, nearest2, &basicNode{q: target}, m2chan)
+		})
+		map1reached := <-m1chan
+		map2reached := <-m2chan
 
 		corners[map1reached] = true
 		corners[map2reached] = true
 
 		_, reachedDelta := planOpts.DistanceFunc(&ConstraintInput{StartInput: map1reached.Q(), EndInput: map2reached.Q()})
-		if reachedDelta < algOpts.JointSolveDist {
+
+		// Second iteration; extend maps 1 and 2 towards the halfway point between where they reached
+		if reachedDelta > algOpts.JointSolveDist {
+			target = referenceframe.InterpolateInputs(map1reached.Q(), map2reached.Q(), 0.5)
+
+			nearest1 := nm.nearestNeighbor(nmContext, planOpts, target, map1)
+			nearest2 := nm.nearestNeighbor(nmContext, planOpts, target, map2)
+
+			utils.PanicCapturingGo(func() {
+				mp.constrainedExtend(ctx, algOpts, rseed1, map1, nearest1, &basicNode{q: target}, m1chan)
+			})
+			utils.PanicCapturingGo(func() {
+				mp.constrainedExtend(ctx, algOpts, rseed2, map2, nearest2, &basicNode{q: target}, m2chan)
+			})
+			map1reached = <-m1chan
+			map2reached = <-m2chan
+			_, reachedDelta = planOpts.DistanceFunc(&ConstraintInput{StartInput: map1reached.Q(), EndInput: map2reached.Q()})
```

Comment:
Does this also need to have the corners map updated for the nodes coming out of `m1chan` and `m2chan` here?

Reply:
Whoops, yes